src/Loader/EvalLoader.php

```diff
@@ -16,24 +16,40 @@
  *
  * @author Joel Wurtz <jwurtz@jolicode.com>
  *
- *  @internal
+ * @internal
  */
-final readonly class EvalLoader implements ClassLoaderInterface
+final class EvalLoader implements ClassLoaderInterface
 {
-    private PrettyPrinterAbstract $printer;
+    /** @var array<class-string, true> */
+    private static array $lockMap = [];
 
     public function __construct(
-        private MapperGenerator $generator,
-        private MetadataFactory $metadataFactory,
+        private readonly MapperGenerator $generator,
+        private readonly MetadataFactory $metadataFactory,
+        private readonly PrettyPrinterAbstract $printer = new Standard(),
     ) {
-        $this->printer = new Standard();
     }
 
     public function loadClass(MapperMetadata $mapperMetadata): void
     {
-        eval($this->printer->prettyPrint($this->generator->generate(
-            $this->metadataFactory->getGeneratorMetadata($mapperMetadata->source, $mapperMetadata->target)
-        )));
+        if (isset(self::$lockMap[$mapperMetadata->className])) {
+            do {
+                usleep(100000); // 0.1 second
+            } while (isset(self::$lockMap[$mapperMetadata->className])); // @phpstan-ignore isset.offset
+
+            if (class_exists($mapperMetadata->className, false)) {
+                return;
+            }
+        }
```

Comment:
This code block looks a little weird, but there is a reason for this.
At first, it looked like this
```php
while (isset(self::$lockMap[$mapperMetadata->className])) {
    usleep(100000); // 0.1 second
}

if (class_exists($mapperMetadata->className, false)) {
    return;
}
```
But there is a problem, class_exists gets called every time, but he only needed if lock existed, so I restructured code to first check lock inside if and then start while loop

Reply:
Not sure you could do that if using the LockFactory, anyway an extra call to class_exists should not be a problem